plugins/times-of-day/api/api.js

```diff
@@ -0,0 +1,187 @@
+var plugin = {},
+	common = require('../../../api/utils/common.js'),
+	plugins = require('../../pluginManager.js'),
+	async = require('async');
+
+(function (plugin) {
+	plugins.register("/i", function (ob) {
+		var params = ob.params;
+
+		if (!params.qstring.begin_session && !params.qstring.events) {
+			return false;
+		}
+
+
+		var appId = params.app_id;
+		var hour = params.qstring.hour;
+		var dow = params.qstring.dow;
+		var events = [];
+
+		if (!appId ||
+			!(parseInt(hour) >= 0 && parseInt(hour) <= 23) ||
+			!(parseInt(dow) >= 0 && parseInt(dow) <= 6)) {
+			return false;
+		}
+
+		var hasSession = false;
+		var hasEvents = false;
+
+		if (params.qstring.begin_session) {
+			hasSession = true;
+		}
+
+		if (params.qstring.events && params.qstring.events.length) {
+			hasEvents = true;
+			events = params.qstring.events;
+		}
+
+		var queryArray = [];
+		var query = {};
+		var criteria = {};
+		var update = {};
+		var options = {};
+
+		if (hasSession) {
+			criteria = {
+				"_id": "tod_" + appId + ":Sessions"
+			};
+
+			var incData = {};
+			incData[dow + "." + hour + ".count"] = 1;
+			var setData = {};
+			setData["_id"] = "tod_" + appId + ":Sessions";
+			setData['app'] = appId;
+
+			update = {
+				$set: setData,
+				$inc: incData
+			};
+
+			options = {
+				upsert: true
+			};
+
+			query = {
+				"criteria": criteria,
+				"update": update,
+				"options": options
+			};
+
+			queryArray.push(query);
+		}
+
+		if (hasEvents) {
+			for (var i = 0; i < events.length; i++) {
+				criteria = {
+					"_id": "tod_" + appId + ":" + events[i].key
+				};
+
+				var incData = {};
+				incData[dow + "." + hour + ".count"] = 1;
+				var setData = {};
+				setData["_id"] = "tod_" + appId + ":" + events[i].key;
+				setData['app'] = appId;
+
+				update = {
+					$set: setData,
+					$inc: incData
+				};
+
+				options = {
+					upsert: true
+				};
+
+				query = {
+					"criteria": criteria,
+					"update": update,
+					"options": options
+				};
+
+				queryArray.push(query);
+			}
+		}
+
+		var parallelTasksArray = [];
+
+		queryArray.forEach((query) => {
+			parallelTasksArray.push(insertUpdateTimesOfDay.bind(null, query));
+		});
+
+		async.parallel(parallelTasksArray, function (err, result) {
+			if (err) {
+				console.log("Error while updating times of day: ", err.message);
+			}
+		})
+
+		function insertUpdateTimesOfDay(query, callback) {
+			common.db.collection('timesofday').update(query.criteria, query.update, query.options, function (err, result) {
```

Comment:
May be as we are constantly updating documents in the loop, it would be better to use bulk processing. In this case all queries are sent to db server and are run there and then only result is returned, consuming only 1 connection, instead of opening new one on each update

```
common.db.onOpened(function(){
    var bulk = common.db._native.collection("timesofday").initializeUnorderedBulkOp();
    for(var i = 0; i < queue.length; i++){
        bulk.find(queue[i].criteria).upsert().updateOne(queue[i].update);
    }
    bulk.execute(function(err, updateResult) {
        if(err){
            //there was an error
        }
        //all done
    });
});
```